A17-client-side-health-checking.md

```diff
@@ -0,0 +1,325 @@
+Checking When Needed" below.)
+
+Note that when an LB policy sees a subchannel go into state
+`TRANSIENT_FAILURE`, it will not be able to tell whether the subchannel
+has become disconnected or whether the backend has reported that it
+is unhealthy.  The LB policy may therefore take some unnecessary but
+unharmful steps in response, such as asking the subchannel to connect
+(which will be a no-op if it's already connected) or asking the resolver
+to re-resolve (which will probably be throttled by the cooldown timer).
+
+#### Client Workflow
+
+When a subchannel first establishes a connection, if health checking
+is enabled, the client will immediately start the `Watch()` call to the
+backend.  The subchannel will stay in state CONNECTING until it receives
+the first health-check response from the backend (see "Race Conditions
+and Subchannel Startup" below).
+
+When a client receives a health-checking response from the backend,
+if the health check response indicates that the backend is healthy, the
+subchannel will transition to state READY; otherwise, it will transition
+to state `TRANSIENT_FAILURE`.  Note that this means that when a backend
+transitions from unhealthy to healthy, the subchannel's connectivity state
+will transition from `TRANSIENT_FAILURE` directly to `READY`, with no stop at
+CONNECTING in between.  This is a new transition that was not [previously
+supported](https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md), although it is unlikely to be a problem for applications,
+because (a) it won't affect the overall state of the parent channel
+unless all subchannels are in the same state at the same time and (b)
+because the connectivity state API can drop states, the application won't
+be able to tell that we didn't stop in `CONNECTING` in between anyway.
+
+If the `Watch()` call fails with status `UNIMPLEMENTED`, the client will
+act as if health checking is disabled.  That is, it will not retry
+the health-checking call, but it will treat the channel as healthy
+(connectivity state `READY`).  However, the client will record a [channel
+trace](https://github.com/grpc/proposal/blob/master/A3-channel-tracing.md)
+event indicating that this has happened.  It will also log a message at
+priority ERROR.
+
+If the `Watch()` call returns any other status, the subchannel will
+transition to connectivity state `TRANSIENT_FAILURE` and will retry the
+call.  To avoid hammering a server that may be experiencing problems,
+the client will use exponential backoff between attempts.  When the client
+receives a message from the server on a given call, the backoff state is
+reset, so the next attempt will occur immediately, but any subsequent
+attempt will be subject to exponential backoff.  When the next attempt
+starts, the subchannel will transition to state `CONNECTING`.
+
+When the client subchannel is shutting down or when the backend sends a
```

Comment:
Since the server never closes the stream, will it stop server from shutting down cleanly? Should the server close all health check streams when shutting down?

Reply:
I think this is an implementation detail.  If the server is shutting down, it will send a GOAWAY, which will cause the client to cancel the stream.  If the server has already proactively cancelled the stream, that won't break anything.